File: frontend/js/services/productoService.js
import API from './api.js';

const ProductoService = {
    async listar() {
        return await API.get('productos');
    },
    
    async buscar(q) {
        return await API.get('productos/buscar', { q });
    },
    
    async buscarPorCodigo(codigo) {
        return await API.get('productos', { codigo });
    },
    
    async bajoStock() {
        return await API.get('productos/bajo-stock');
    },
    
    async guardar(producto) {
        // Validaciones en cliente antes de mandar
        if (!producto.codigo || producto.codigo.trim() === '') throw new Error('El código es obligatorio.');
        if (!producto.nombre || producto.nombre.trim() === '') throw new Error('El nombre es obligatorio.');
        if (parseFloat(producto.precioVenta) <= 0) throw new Error('El precio de venta debe ser mayor a cero.');
        if (parseInt(producto.stock) < 0) throw new Error('El stock no puede ser negativo.');
        
        if (producto.idProducto) {
            return await API.put('productos', producto);
        } else {
            return await API.post('productos', producto);
        }
    },
    
    async eliminar(id) {
        return await API.delete('productos', { id });
    },
    
    // Categorías y Proveedores asociados a la gestión de productos
    async listarCategorias() {
        return await API.get('categorias');
    },
    
    async listarProveedores() {
        return await API.get('proveedores');
    }
};

export default ProductoService;

File: backend/src/main/java/com/pukio/service/ProductoService.java
package com.pukio.service;

import com.pukio.dao.ProductoDAO;
import com.pukio.model.Producto;
import java.sql.SQLException;
import java.util.List;

public class ProductoService {

    private final ProductoDAO dao = new ProductoDAO();

    public List<Producto> listar() throws SQLException { return dao.listarActivos(); }

    public List<Producto> buscarPorNombre(String q) throws SQLException { return dao.buscarPorNombre(q); }

    public Producto buscarPorCodigo(String c) throws SQLException { return dao.buscarPorCodigo(c); }

    public List<Producto> bajoStock() throws SQLException { return dao.listarBajoStock(); }

    public void guardar(Producto p) throws Exception {
        validar(p);
        if (p.getIdProducto() == 0) dao.insertar(p);
        else dao.actualizar(p);
    }

    public void eliminar(int id) throws SQLException { dao.eliminar(id); }

    private void validar(Producto p) throws Exception {
        if (p.getCodigo() == null || p.getCodigo().isBlank()) throw new Exception("El codigo es obligatorio.");
        if (p.getNombre() == null || p.getNombre().isBlank()) throw new Exception("El nombre es obligatorio.");
        if (p.getPrecioVenta() <= 0) throw new Exception("El precio de venta debe ser mayor a 0.");
        if (p.getStock() < 0) throw new Exception("El stock no puede ser negativo.");
    }
}
